Add support for access notes

beccadax opened the pull request:
Access notes are sidecar files that can be ingested while a module is being compiled to apply small tweaks to the declarations in the module. They're intended to be used in situations where a module needs to be built with additional internals exposed, but without editing its source code. To make sure the differences from the source code are obvious, the compiler diagnoses each change made by the access notes.

Currently, access notes can:

* Add or remove `@objc` attributes
* Set the selector for an `@objc` attribute
* Add or remove `dynamic` modifiers

In the future, they might gain other capabilities, such as increasing access levels, adding SPI groups, or enabling library evolution for a module.

Fixes rdar://71870054, rdar://71872830.

------

At the time of this PR's creation, this work is incomplete, but basically functions. I still need to write a more targeted set of tests, but I want to self-review the PR in its current state so I can identify other tasks I have to complete.

beccadax committed:
Apply access notes

beccadax committed:
Warn about modifications made by access notes

beccadax committed:
Add ability to load access notes

beccadax committed:
Flatten access note format

Rather than nesting notes for members inside a note for the type, access note files now contain a flat list of notes with parent names embedded in the `Name` property.

Fixes rdar://71872830.

fixup spacing in access notes

beccadax committed:
Clarify behavior of access note with ObjCName only

An access note with an ObjCName property but no ObjC property now implicitly sets ObjC to true. ObjC: false with an ObjCName is still illegal.

beccadax committed:
Let access notes apply to accessors

`Name: ‘getter:Context.property()’` now applies an access note to a getter of Context.property, and `Name: ‘setter:Context.property(_:)’` applies one to its setter. Additionally, access notes without `getter:` or `setter:` now cannot apply to accessors.

beccadax committed:
[NFC] Make access note diagnostic locations more consistent

Access note diagnostics tend to need a fairly specific “diagnose at attr location if possible, decl location otherwise” pattern; encapsulate that in a helper function and use it everywhere.

beccadax commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -5579,5 +5579,36 @@ ERROR(atomics_ordering_must_be_constant, none,
+// MARK: access notes
+//------------------------------------------------------------------------------
+
+WARNING(attr_added_by_access_note, none,
```
These probably ought to be remarks, not warnings, to prevent -warnings-as-errors from kicking in.

beccadax commented on include/swift/Option/FrontendOptions.td:
```diff
@@ -33,6 +33,9 @@ def supplementary_output_file_map : Separate<["-"], "supplementary-output-file-m
 def frontend_parseable_output : Flag<["-"], "frontend-parseable-output">,
   HelpText<"Emit textual output in a parseable format">;
 
+def access_notes : Separate<["-"], "access-notes">, MetaVarName<"<path>">,
```
This should probably be `-access-notes-path` or something to match the flags for other compiler inputs. I think there might also be a way I should be marking this flag as having a path argument?

beccadax commented on lib/AST/AccessNotes.cpp:
```diff
@@ -0,0 +1,295 @@
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/YAMLTraits.h"
+
+// FIXME: Copied from MiscDiagnostics--don't do that.
+static llvm::Optional<swift::ObjCSelector>
+parseObjCSelector(swift::ASTContext &ctx, llvm::StringRef string) {
```
Should clean up this duplication.

beccadax commented on lib/AST/AccessNotes.cpp:
```diff
@@ -0,0 +1,295 @@
+}
+
+bool AccessNoteDeclName::matches(ValueDecl *VD) const {
+  // These are normally just `VD` and `name`, but not for accessors.
```
Oops, this comment needs to be corrected.

beccadax commented on lib/AST/AccessNotes.cpp:
```diff
@@ -0,0 +1,295 @@
+  for (const auto &note : notes) {
+    os << "\n";
+    note.dump(os, /*indent=*/2);
+  }
```
Needs a closing ")", doesn't it?

beccadax commented on include/swift/AST/AccessNotes.h:
```diff
@@ -0,0 +1,77 @@
+
+class AccessNote {
+public:
+  AccessNoteDeclName name;
```
There's no real reason for `name` to be lowercase but the others to be uppercase—they all correspond to uppercased fields in the YAML file.

beccadax commented on include/swift/AST/AccessNotes.h:
```diff
@@ -0,0 +1,77 @@
+class AccessNoteDeclName {
+public:
+  std::vector<Identifier> parentNames;
+  DeclName name;
```
Is there something better than `name.name` here?

beccadax commented on include/swift/AST/AccessNotes.h:
```diff
@@ -0,0 +1,77 @@
+
+class AccessNotes {
+public:
+  std::string reason;
+  std::vector<AccessNote> notes;
```
Likewise on capitalization.

beccadax commented on lib/AST/AccessNotes.cpp:
```diff
@@ -0,0 +1,295 @@
+  io.mapRequired("Name", note.name);
+  io.mapOptional("ObjC", note.ObjC);
+  io.mapOptional("Dynamic", note.Dynamic);
+  io.mapOptional("ObjCName", note.ObjCName);
```
Can we make this ignore unknown keys instead of erroring out?

beccadax replied:
Without modifying `llvm::yaml::Input`, the best we can do is mark unused keys as used (to suppress the error) and add them to a list to be diagnosed.

beccadax commented on lib/AST/AccessNotes.cpp:
```diff
@@ -0,0 +1,295 @@
+
+void MappingTraits<AccessNotes>::mapping(IO &io, AccessNotes &notes) {
+  io.mapRequired("Reason", notes.reason);
+  io.mapRequired("Notes", notes.notes);
```
Again, warning instead of error?

beccadax commented on lib/Frontend/Frontend.cpp:
```diff
@@ -914,6 +915,21 @@ ModuleDecl *CompilerInstance::getMainModule() const {
+      auto bufferOrError =
+          swift::vfs::getFileOrSTDIN(getFileSystem(),
+                               Invocation.getFrontendOptions().AccessNotesPath);
+      // FIXME: Diagnose properly
+      auto buffer = cantFail(llvm::errorOrToExpected(std::move(bufferOrError)),
+                             "can't open access notes file");
+
+      auto expectedAccessNotes = AccessNotes::load(*Context, buffer.get());
+      // FIXME: Diagnose properly
+      MainModule->getAccessNotes() = cantFail(std::move(expectedAccessNotes),
+                                              "invalid access notes:");
```
Need real error handling (probably warnings)

beccadax commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1385,6 +1386,110 @@ void TypeChecker::diagnoseDuplicateCaptureVars(CaptureListExpr *expr) {
+                            const AccessNotes &notes) {
+  ASTContext &ctx = VD->getASTContext();
+
+  addOrRemoveAttr<ObjCAttr>(VD, notes, note.ObjC, [&]() {
+    return ObjCAttr::create(ctx, note.ObjCName, false);
+  });
+
+  addOrRemoveAttr<DynamicAttr>(VD, notes, note.Dynamic, [&]{
```
Nit: Silly that these lambdas use slightly different syntax.

beccadax commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1385,6 +1386,110 @@ void TypeChecker::diagnoseDuplicateCaptureVars(CaptureListExpr *expr) {
 }
 
+template<typename ...DiagIDAndArgs> InFlightDiagnostic
+diagnoseForAccessNote(DeclAttribute *attr, ValueDecl *VD,
```
Nit: `diagnoseAtAttrOrDecl()`?

beccadax commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1385,6 +1386,110 @@ void TypeChecker::diagnoseDuplicateCaptureVars(CaptureListExpr *expr) {
+evaluator::SideEffect
+ApplyAccessNoteRequest::evaluate(Evaluator &evaluator, ValueDecl *VD) const {
+  AccessNotes &notes = VD->getModuleContext()->getAccessNotes();
+  if (auto note = notes.lookup(VD))
```
Applying access notes lazily like this is mostly great, except that it becomes impossible to diagnose that an access note doesn't match anything. It might be better to walk through the access notes after parsing, looking up the declarations matching their names and applying them ahead of time.

beccadax replied:
This needs more thought, so I'm adding a FIXME to the test and otherwise leaving it for later.

tschuett commented on lib/AST/AccessNotes.cpp:
```diff
@@ -0,0 +1,295 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
The copyright years are off.

beccadax replied:
Thanks!

beccadax committed:
[NFC] Correct copyright years

beccadax committed:
Correct access notes flag spelling and add to driver

beccadax committed:
Handle access note I/O and parse errors

beccadax committed:
[NFC] Centralize ObjCSelector helper

beccadax committed:
Allow unknown keys in access notes

LLVM’s YAML support wants to show a hard error if an unknown string appears in an access note. Instead, emit a remark but load the parts of the access note we do understand to allow for future expansion of access notes.

beccadax committed:
[NFC] Clean up access note code

• Rename properties for greater consistency with the YAML file
• Rename AccessNotes -> AccessNotesFile
• Add doc comments
• Miscellaneous comment/style improvements.

beccadax committed:
[NFC] Note missing access note validation

beccadax committed:
Diagnose @objc(selector) access note conflicts

beccadax committed:
Weaken access note warnings into remarks

beccadax commented:
@swift-ci please test

beccadax committed:
[NFC] Fix Windows driver tests

beccadax committed:
Update access note serialization for LLVM rebranch

beccadax committed:
Improve access note error handling

The LLVM rebranch added an “AllowUnknownKeys” setting to llvm::yaml::Input, which lets us rip out a lot of marginal code and encourages a broader rework of access note error diagnosis:

• Access note warnings and errors are now diagnosed as they’re found during YAML parsing
• They now have proper SourceLocs inside the .accessnotes file
• They’re now tested using -verify-additional-file instead of FileCheck
• A lot of gross duct tape is now gone

beccadax commented:
Needed a rebase on top of rebranch, but llvm::yaml has vastly improved in the new LLVM, which made error handling a lot nicer.

@swift-ci please test

beccadax committed:
Fix attempts to fix Windows tests

beccadax commented:
@swift-ci please test

beccadax committed:
[NFC] Handle Windows error spelling difference

lolsob

beccadax commented:
The Linux failure was an unrelated transient; the Windows one was literally that they spell an error message with "no" instead of "No".

@swift-ci please test

beccadax commented:
@swift-ci please test macOS platform

xymus commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -2311,6 +2311,25 @@ class SpecializeAttrTargetDeclRequest
+/// Looks up and applies the access note for a given declaration.
+class ApplyAccessNoteRequest
+    : public SimpleRequest<ApplyAccessNoteRequest,
+                           evaluator::SideEffect(ValueDecl *),
```
Do you need `SideEffect` for a single return value?

xymus replied:
Ah, because the result value is dropped at the call sites I guess? I understand the use of `SideEffect` in this case.

beccadax replied:
Yeah. `ApplyAccessNoteRequest` mutates the AST; I'm basically using the request evaluator here to avoid applying the access notes twice or reentering access note application for the same declaration.

xymus commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1445,6 +1446,110 @@ void TypeChecker::diagnoseDuplicateCaptureVars(CaptureListExpr *expr) {
+        note.fixItReplace({ attr->getLParenLoc(), attr->getRParenLoc() },
+                          newNameString);
+      else
+        note.fixItInsertAfter(attr->getLocation(), newNameString);
```
Is this worked expected to be done during normal compilation? If so, do we expect the devs to apply these for the source to fit with the access notes?

beccadax replied:
In some workflows, yes, we do expect devs to apply these fix-its.

xymus reviewed (approved):
This looks good and I like the straightforward design.

beccadax commented:
@swift-ci please test macOS platform
